File: caribou/deployment_solver/models/indexer.py
from abc import ABC
from typing import Any


class Indexer(ABC):
    def __init__(self) -> None:
        if not hasattr(self, "_value_indices"):
            self._value_indices: dict[str, int] = {}

        self._indices_to_values: dict[int, str] = self.indicies_to_values()

    def get_value_indices(self) -> dict[str, int]:
        return self._value_indices

    def indicies_to_values(self) -> dict[int, str]:
        return {index: instance for instance, index in self._value_indices.items()}

    def value_to_index(self, value: str) -> int:
        return self._value_indices[value]

    def index_to_value(self, index: int) -> str:
        return self._indices_to_values[index]

    def to_dict(self) -> dict[str, Any]:
        return {"value_indices": self._value_indices, "indices_to_values": self._indices_to_values}

File: caribou-go/src/models/indexer.go
package models

import (
	"strconv"
)

type Indexer struct {
	valueIndices  map[string]int
	indicesValues map[int]string
}

func (in *Indexer) getValueIndices() map[string]int {
	return in.valueIndices
}

func indicesToValues(valueIndices map[string]int) map[int]string {
	var result map[int]string
	for instance, index := range valueIndices {
		result[index] = instance
	}
	return result
}

func (in *Indexer) ValueToIndex(value string) int {
	return in.valueIndices[value]
}

func (in *Indexer) IndexToValue(index int) string {
	return in.indicesValues[index]
}

type RegionIndexer struct {
	Indexer
}

func NewRegionIndexer(regions []interface{}) *RegionIndexer {
	var valueIndices map[string]int
	for index, region := range regions {
		valueIndices[region.(string)] = index
	}
	re := RegionIndexer{Indexer{
		valueIndices:  valueIndices,
		indicesValues: indicesToValues(valueIndices),
	}}
	return &re
}

type InstanceIndexer struct {
	Indexer
}

func NewInstanceIndexer(nodes []interface{}) *InstanceIndexer {
	var valueIndices map[string]int
	for index, node := range nodes {
		nodeMap := node.(map[string]interface{})
		valueIndices[nodeMap["instance_name"].(string)] = index
	}
	re := InstanceIndexer{Indexer{
		valueIndices:  valueIndices,
		indicesValues: indicesToValues(valueIndices),
	}}
	return &re
}

func parseIndexer(indexerData map[string]interface{}) Indexer {
	valueIndices := map[string]int{}
	for k, v := range indexerData["value_indices"].(map[string]interface{}) {
		valueIndices[k] = int(v.(float64))
	}
	indicesValues := map[int]string{}
	for k, v := range indexerData["indices_to_values"].(map[string]interface{}) {
		kInt, _ := strconv.Atoi(k)
		indicesValues[kInt] = v.(string)
	}
	return Indexer{
		valueIndices:  valueIndices,
		indicesValues: indicesValues,
	}
}

func CopyInstanceIndexer(indexerData map[string]interface{}) *InstanceIndexer {
	return &InstanceIndexer{
		parseIndexer(indexerData),
	}
}

func CopyRegionIndexer(indexerData map[string]interface{}) *RegionIndexer {
	return &RegionIndexer{
		parseIndexer(indexerData),
	}
}
